Add support for JVM based projects

arthurscchan opened the pull request:
This PR attempts to add support in the base logic for JVM based projects.

arthurscchan commented:
This is still under process. Not ready to review yet.

arthurscchan commented:
To run it:
```
python3 -m data_prep.introspector <PROJECT_NAME_IN_OSS_D -m 5 -o <RESULT_DIR>
export LLM_NUM_EXP=<NUMBER_OF_PARALLEL_THREAD>
./run_all_experiments.py --model=gpt-3.5-turbo --benchmarks-directory=<RESULT_DIR_FROM_ABOVE>
```

oliverchang reviewed (commented):
This is awesome! 

Can you please add some java benchmarks to `benchmark-sets/java`?  

Then @DavidKorczynski can help test this via gcbrun:

```
/ gcbrun request_pr_exp.py -n test-java -b java
```

(Ignore the space after the / above. trying to avoid actually running it).

oliverchang commented on data_prep/introspector.py:
```diff
@@ -343,10 +359,19 @@ def populate_benchmarks_using_introspector(project: str, language: str,
-  result = project_src.search_source(project, filenames)
+  if language == 'jvm':
+    filenames = [
+        "%s.java" %
```
can use you single quotes for all strings in your PR?

arthurscchan replied:
Changed.

oliverchang commented on data_prep/introspector.py:
```diff
@@ -343,10 +359,19 @@ def populate_benchmarks_using_introspector(project: str, language: str,
-  result = project_src.search_source(project, filenames)
+  if language == 'jvm':
+    filenames = [
+        "%s.java" %
```
Can we just do a string concatenation (or use f-string) here?

arthurscchan replied:
Changed.

oliverchang commented on experiment/textcov.py:
```diff
@@ -193,8 +273,36 @@ def subtract_covered_lines(self, other: Textcov):
     return sum(f.covered_lines for f in self.functions.values())
+
+  def determine_jvm_arguments_type(self, desc: str) -> List[str]:
+    """Determine list of jvm arguments type"""
```
Can you give some more detail on what `desc` is here in the docstring, and what exactly this function is doing?

arthurscchan replied:
Given.

oliverchang commented on data_prep/project_src.py:
```diff
@@ -291,22 +298,29 @@ def _identify_fuzz_targets(
+          interesting_filepaths.append(path)
+        if path.endswith('.java'):
+          potential_harnesses.append(path)
+      else:
```
nit: add a comment to describe what the else branch is.

```python
else:
  # C/C++
```

oliverchang commented on data_prep/introspector.py:
```diff
@@ -296,10 +296,23 @@ def _get_clean_arg_types(function: dict, project: str) -> list[str]:
+  if language == 'jvm':
+    jvm_args = _get_clean_arg_types(function, project)
+    arg_names = [
+        'var_%s' % (name.split('.')[-1].replace("[]", "")) for name in jvm_args
```
Can you explain a bit more about what this is doing and why we are prepending the names with `var_`? Is this because of some limitation with getting the original argument names with how introspector works?

arthurscchan replied:
The fuzz-introspector front end of JVM projects cannot get the original argument name. Thus the argument name here uses var_{argument_type} as argument name reference. Some argument types are full-qualified names of Java classes with [] and . and that is not allowed for Java variable names and they are removed.

oliverchang replied:
Thanks for explaining. Can you please include this detail as a code comment?

arthurscchan replied:
have already done so.

DavidKorczynski commented:
/gcbrun request_pr_exp.py -n test-gemini-1-5 -m vertex_ai_gemini-1-5

oliverchang commented:
/gcbrun request_pr_exp.py -n test-jvm -m vertex_ai_gemini-1-5 -b jvm

oliverchang commented:
I created #274 from this to run the gcbrun experiments. They currently only work on branches on this repo, and not forked repos.

oliverchang commented:
@arthurscchan are you able to generate a successful report  based on this? 

From https://github.com/google/oss-fuzz-gen/pull/274, we get https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-05-15-274-test-jvm-jvm/index.html which is not  showing  any results.

arthurscchan commented:
@oliverchang I did not have access to the sample report. But I do think that I did not specifically change things for the report generation process. I am mostly focus on the flow and coverage report diff. Thus I am not sure if there are changes needed for the JVM report and I am not even aware of the part of code for the report generation. Can you point me to the code location for the report generation and I could go and double confirm that? Sorry for the overseeing of the report part.

DavidKorczynski commented:
> @arthurscchan are you able to generate a successful report based on this?
> 
> From #274, we get https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-05-15-274-test-jvm-jvm/index.html which is not showing any results.

Column `Status` shows them all as *running*, is this accurate? The results doesn't match what we get from local runs

arthurscchan commented:
@oliverchang 

For reference, here is the result I got from my local run of those 5 benchmarks.
```
================================================================================                                                                                                       
*fuzzywuzzy, [me.xdrop.fuzzywuzzy.algorithms.DefaultStringProcessor].process(java.lang.String)*
build success rate: 1.0, crash rate: 0.0, found bug: 0, max coverage: 0.025390625, max line coverage diff: 2.75
max coverage sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-fuzzywuzzy-me.xdrop.fuzzywuzzy.algorithms.defaultstringprocessor.process_java.lang.string/fixed_targets/01.java
max coverage diff sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-fuzzywuzzy-me.xdrop.fuzzywuzzy.algorithms.defaultstringprocessor.process_java.lang.string/fixed_targets/01.ja
va                                                                                                                                                                                     
max coverage diff report: None                                                                                                                                                         
                                                                                                                                                                                       
================================================================================
*fuzzywuzzy, [me.xdrop.diffutils.DiffUtils].levEditDistance(java.lang.String,java.lang.String,int)*
build success rate: 1.0, crash rate: 0.0, found bug: 0, max coverage: 0.130859375, max line coverage diff: 25.5
max coverage sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-fuzzywuzzy-me.xdrop.diffutils.diffutils.leveditdistance_java.lang.string_java.lang.string_int/fixed_targets/02.jav
a                                                                                                                                                                                      
max coverage diff sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-fuzzywuzzy-me.xdrop.diffutils.diffutils.leveditdistance_java.lang.string_java.lang.string_int/fixed_targets/0
2.java                                                                                                                                                                                 
max coverage diff report: None

================================================================================
*sqlite-jdbc, [org.sqlite.date.DateFormatUtils].format(long,java.lang.String)*
build success rate: 1.0, crash rate: 0.0, found bug: 0, max coverage: 0.75, max line coverage diff: 3.782608695652174
max coverage sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-sqlite-jdbc-org.sqlite.date.dateformatutils.format_long_java.lang.string/fixed_targets/01.java
max coverage diff sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-sqlite-jdbc-org.sqlite.date.dateformatutils.format_long_java.lang.string/fixed_targets/01.java
max coverage diff report: None

================================================================================                                                                                                       
*sqlite-jdbc, [org.sqlite.date.DateFormatUtils].formatUTC(long,java.lang.String,java.util.Locale)*
build success rate: 1.0, crash rate: 0.5, found bug: 1, max coverage: 0.7421875, max line coverage diff: 3.765217391304348
max coverage sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-sqlite-jdbc-org.sqlite.date.dateformatutils.formatutc_long_java.lang.string_java.util.locale/fixed_targets/01.java
max coverage diff sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-sqlite-jdbc-org.sqlite.date.dateformatutils.formatutc_long_java.lang.string_java.util.locale/fixed_targets/01
.java
max coverage diff report: None

================================================================================
*jsemver, [com.github.zafarkhaja.semver.Version].tryParse(java.lang.String,boolean)*
build success rate: 1.0, crash rate: 0.0, found bug: 0, max coverage: 0.47265625, max line coverage diff: 56.666666666666664
max coverage sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-jsemver-com.github.zafarkhaja.semver.version.tryparse_java.lang.string_boolean/fixed_targets/02.java
max coverage diff sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-jsemver-com.github.zafarkhaja.semver.version.tryparse_java.lang.string_boolean/fixed_targets/02.java
max coverage diff report: None

================================================================================
*jsemver, [com.github.zafarkhaja.semver.Version].of(long,long,long)*
build success rate: 1.0, crash rate: 0.0, found bug: 0, max coverage: 0.04296875, max line coverage diff: 3.0
max coverage sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-jsemver-com.github.zafarkhaja.semver.version.of_long_long_long/fixed_targets/01.java
max coverage diff sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-jsemver-com.github.zafarkhaja.semver.version.of_long_long_long/fixed_targets/01.java
max coverage diff report: None

================================================================================
*json-java, [org.json.JSONArray].putAll(java.lang.String[])*
build success rate: 0.0, crash rate: 0.0, found bug: 0, max coverage: 0.0, max line coverage diff: 0.0
max coverage sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-json-java-org.json.jsonarray.putall_java.lang.string/fixed_targets/01.java
max coverage diff sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-json-java-org.json.jsonarray.putall_java.lang.string/fixed_targets/01.java
max coverage diff report: None

================================================================================
*json-java, [org.json.JSONArray].putAll(int[])*
build success rate: 1.0, crash rate: 0.0, found bug: 0, max coverage: 0.056640625, max line coverage diff: 8.25
max coverage sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-json-java-org.json.jsonarray.putall_int/fixed_targets/01.java
max coverage diff sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-json-java-org.json.jsonarray.putall_int/fixed_targets/01.java
max coverage diff report: None

================================================================================
*twitter4j, [twitter4j.OAuth2Token].equals(java.lang.Object)*
build success rate: 0.0, crash rate: 0.0, found bug: 0, max coverage: 0.0, max line coverage diff: 0.0
max coverage sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-twitter4j-twitter4j.oauth2token.equals_java.lang.object/fixed_targets/01.java
max coverage diff sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-twitter4j-twitter4j.oauth2token.equals_java.lang.object/fixed_targets/01.java
max coverage diff report: None

================================================================================
*twitter4j, [twitter4j.TimeSpanConverter].toTimeSpanString(long)*
build success rate: 0.0, crash rate: 0.0, found bug: 0, max coverage: 0.0, max line coverage diff: 0.0
max coverage sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-twitter4j-twitter4j.timespanconverter.totimespanstring_long/fixed_targets/01.java
max coverage diff sample: /mnt/volume_lon1_01/oss-fuzz-gen/results/output-twitter4j-twitter4j.timespanconverter.totimespanstring_long/fixed_targets/01.java
max coverage diff report: None
```

oliverchang commented:
> > @arthurscchan are you able to generate a successful report based on this?
> > From #274, we get https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-05-15-274-test-jvm-jvm/index.html which is not showing any results.
> 
> Column `Status` shows them all as _running_, is this accurate? The results doesn't match what we get from local runs

They seem to be erroring out , so they're left as "Running".

oliverchang commented:
> > > @arthurscchan are you able to generate a successful report based on this?
> > > From #274, we get https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-05-15-274-test-jvm-jvm/index.html which is not showing any results.
> > 
> > 
> > Column `Status` shows them all as _running_, is this accurate? The results doesn't match what we get from local runs
> 
> They seem to be erroring out , so they're left as "Running".

This is because our web report generation didn't work for Java. Fixed in https://github.com/google/oss-fuzz-gen/pull/275.

oliverchang commented:
With the report fix in and https://github.com/google/oss-fuzz-gen/commit/8c4e5ba264f1406c50d06966254c4083cd69aa88, we now get a working web report:

https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-05-16-274-test-jvm-4-jvm/index.html#

The line cov diff shows as all 0 though, which is inconsistent to https://github.com/google/oss-fuzz-gen/pull/236#issuecomment-2113109959. Can you please take a look to see if anything is wrong, or if it's just because the quality of the targets generated are lower?

oliverchang commented on experiment/textcov.py:
```diff
@@ -172,6 +202,56 @@ def from_file(
     return textcov
 
+  @classmethod
+  def from_jvm_file(cls, file_path) -> Textcov:
```
Can we be consistent here, and pass a `file_handle` instead?

oliverchang replied:
Question: does this ignore coverage in the fuzz target source?

oliverchang commented on experiment/evaluator.py:
```diff
@@ -85,12 +86,18 @@ def load_existing_textcov(project: str) -> textcov.Textcov:
-    if not blob.name.endswith('.covreport'):
-      continue
+    if language == 'jvm':
+      if not blob.name == 'jacoco.xml':
```
nit: just `if blob.name !=`.

oliverchang commented:
The latest report  with  some fixes incoporated (from https://github.com/google/oss-fuzz-gen/pull/274/) is https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-05-16-274-test-jvm-5-jvm/index.html

The percentages look wrong -- we have one that claims a 5666.67% coverage increase. 

This looks consistent with the same benchmark in https://github.com/google/oss-fuzz-gen/pull/236#issuecomment-2113109959 where  "56.666666666666664" is reported.

The number there should be a number in the range [0.0, 1.0], so something must be going wrong the the division here: https://github.com/google/oss-fuzz-gen/blob/b1a654007b13c5129d8c8e5b7cfc7a1341f4254d/experiment/evaluator.py#L344

@arthurscchan can you please take a look?

arthurscchan commented on experiment/builder_runner.py:
```diff
@@ -119,20 +119,42 @@ def _get_minimum_func_name(self, func_sig: str) -> str:
+    with open(target_path) as generated_code_file:
+      code = generated_code_file.read()
+
+    base_arg_regex = r'[a-zA-Z_$][a-zA-Z_$0-9(),.]*'
```
Added.

arthurscchan commented:
@oliverchang I have added comments and cherry picking the changes from #274 accordingly. I do found the reason for the strange max coverage data. It is mainly because of the different information in the summary.json for the coverage report of JVM projects. It lacks the information of total lines in the project and thus the total lines information used for the calculation is wrong. I have fixed the logic by directly getting the total line count from the jacoco.xml report from the current one in order to retrieve the correct number for the calculation. To answer your question, the number used already deducted the lines that has been covered in existing coverage report, thus the number shown in the result is what has been gained from the new fuzzers generated by oss-fuzz-gen.

arthurscchan commented:
@oliverchang Additional information, the summary.json for the coverage report of JVM projects lacks that information. And after some search, it seems that sumamry.json is generated from the oss-fuzz side in https://github.com/google/oss-fuzz/blob/master/infra/base-images/base-runner/coverage#L218-L227 with the script  jacoco_report_converter.py. Thus if we need to make the summary.json have the needed information, we will need to change the infer of oss-fuzz to accommodate that. I think at the moment, using my approach in oss-fuzz-gen to retrieve the total line number by parsing the jacoco.xml of the current run is a better option to get the correct max line coverage diff number for JVM projects.

oliverchang commented:
Thanks! The latest report is now available here: https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-05-17-274-test-jvm-6-jvm/index.html
